src/instructlab/sdg/utils/docprocessor.py

```diff
@@ -0,0 +1,595 @@
+                    book_title = book_text
+                    book_text = f"# Title: **{book_text}**"
+                else:
+                    book_text = f"## **{book_text}**"
+
+            if book_element["type"] == "title":
+                book_text = f"# **{book_text}**"
+            if book_element["type"] == "page-header":
+                book_text = f"Page Header: **{book_text}**\n\n"
+
+            if chunking_criteria is not None:
+                # custom break function that can be used to chunk document
+                if chunking_criteria(book_text):
+                    document_chunks.append("\n\n".join(current_buffer))
+                    current_buffer = []
+            elif (
+                prev_page_number is not None
+                and prev_page_number != current_book_page_number
+            ) and keep_same_page_thing_together:
+                document_chunks.append("\n\n".join(current_buffer))
+                current_buffer = []
+            else:
+                if (
+                    get_token_count("\n\n".join(current_buffer), tokenizer)
+                    >= max_token_per_chunk
+                    and len(current_buffer) > 1
+                ):
+                    document_chunks.append("\n\n".join(current_buffer[:-1]))
+
+                    if (
+                        get_token_count(current_buffer[-1], tokenizer)
+                        >= max_token_per_chunk
+                    ):
+                        document_chunks.append(current_buffer[-1])
+                        current_buffer = []
+                    else:
+                        current_buffer = current_buffer[-1:]
+
+            if book_element["type"] == "paragraph":
+                book_text = add_heading_formatting(book_text)
+            elif book_element["type"] == "table":
+                book_text = get_table(json_book, book_element["$ref"])
+
+            if "## References" in book_text or "## Acknowledgements" in book_text:
+                # For research papers we ignore everything after these sections
+                break
+            current_buffer.append(book_text)
+
+        try:
+            prev_page_number = current_book_page_number
+        except Exception as e:
+            logger.error(f"Error processing book element: {book_element}, {str(e)}")
+
+    if "\n\n".join(current_buffer) not in document_chunks:
+        document_chunks.append("\n\n".join(current_buffer))
+    return document_chunks
+
+
+def safe_concatenate_datasets(datasets: list):
+    """
+    Concatenate datasets safely, ignoring any datasets that are None or empty.
+    """
+    filtered_datasets = [ds for ds in datasets if ds is not None and ds.num_rows > 0]
+
+    if not filtered_datasets:
+        return None
+
+    return concatenate_datasets(filtered_datasets)
+
+
+class DocProcessor:
+    def __init__(
+        self,
+        parsed_doc_dir: Path,
+        tokenizer_model_name: str = "TheBloke/Mixtral-8x7B-Instruct-v0.1-GGUF",
```

Comment:
This should come from the model_name parameter passed to generate_data.py's `generate_data`. That way we load the tokenizer for the same teacher model that's being used.